src/main.rs

```diff
@@ -73,119 +76,32 @@ fn run_in_userns() -> anyhow::Result<()> {
     // this parent process sets up user namespace mappings, notifies the child to continue,
     // then waits for the child to exit
     close(reader)?;
-    setup_id_maps(child, user_id, group_id).context("Failed to setup id mappings")?;
+    // Kill the child process if we fail to setup the uid/gid mappings
+    if let Err(e) =
+        setup_id_maps(child, user_id, group_id).context("Failed to setup uid/gid mappings")
+    {
+        signal::kill(child, Signal::SIGTERM)?;
+        return Err(e);
```

Comment:
I suspect you want to `wait` on the child after doing this, probably doesn't make any meaningful difference (maybe running under `tini` might be strange?) but leaking children is a slightly unusual thing to do if you can help it.

Reply:
So the child is not actually always present, which is what is delaying me from progressing this PR further right now as I don't understand what is happening in that case.

Reply:
Presumably you get to this branch by not having enough subuids?

Reply:
Ok, I've found a couple of bugs which should unblock you:

1. You are calling `read(reader, &mut Vec::new())` which leads to a syscall of `read(3, <ptr>, 0)` (i.e. a read into a zero-length buffer), but there's special case handling for this:

    >If count is zero, read() may detect the errors described below. In the absence of any errors, or if read() does not check for errors, a read() with a count of 0 returns zero and has no other effects.

    So you should pass a `&mut [0u8; 1]` (or similar) to cause the read to actually block for data/EOF.  Currently the child process is just starting in parallel with the parent (before the uid-mappings are in place) and often in testing gets to the point of calling execve before the mappings which might cause the child to try and do rootless logic itself.

2. This might have been my fault.  Reading the manpage for `wait/waitpid` (e.g. https://man7.org/linux/man-pages/man2/wait.2.html, specifically the NOTES section) says:

    > POSIX.1-2001 specifies that if the disposition of SIGCHLD is set
    >  to SIG_IGN or the SA_NOCLDWAIT flag is set for SIGCHLD (see
    >  [sigaction(2)](https://man7.org/linux/man-pages/man2/sigaction.2.html)), then children that terminate do not become zombies
    >  and a call to wait() or waitpid() will block until all children
    >  have terminated, and then fail with [errno](https://man7.org/linux/man-pages/man3/errno.3.html) set to ECHILD.  (The
    >  original POSIX standard left the behavior of setting SIGCHLD to
    >  SIG_IGN unspecified.  Note that even though the default
    >  disposition of SIGCHLD is "ignore", explicitly setting the
    >  disposition to SIG_IGN results in different treatment of zombie
    >  process children.)

    In your case that's not quite what's happening but it gave me the hint I needed: you don't have `SIGCHLD` masked or ignored, _but_ you aren't registering for a "child death signal" in your `clone` call.   Because of this libc/kernel isn't adding the child to it's "list of children I might wait for" and hence the `wait` response with the "you have no children to wait on" error code.

    Passing `Some(SIGCHLD as c_int)` as the signal makes everything "work" though I'm now hitting some permissions errors trying to open system-wide logs?

Reply:
Thanks, this is now working locally so I'll look into adding CI tests